### PR TITLE
Delete infra container if network plugin fails

### LINKED ISSUE #14938: Pod still created if network plugin returns error
I'm seeing a bug where the network plugin fails to configure networking, and yet the containers for this pod are still deployed (without networking configured).

It looks like the following is occurring:
1) Kubelet attempts to create infra container in `SyncPod`
2) Network plugin fails, returns error code.
3) Kubelet detects error and skips pod creation but does not tear down the infra container
4) On the next `SyncPod` call, the kubelet detects that the infra container is running and deploys the other containers in the pod.

The fix here is for the kubelet to clean up the infra container in `SyncPod` when it receives the error from the network plugin and before it returns.

Here are some logs illustrating the steps I described above:
```
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: I1001 14:24:07.079239   11828 exec.go:137] SetUpPod 'exec' network plugin output: , exit status 1
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: E1001 14:24:07.079680   11828 manager.go:1735] Failed to create pod infra container: exit status 1; Skipping pod "busybox_default"
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: I1001 14:24:07.079928   11828 kubelet.go:2529] Generating status for "busybox_default"
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: I1001 14:24:07.088250   11828 manager.go:309] Container inspect result: {ID:8ff588a7a7559041d43ed570b800304d29284573b3c8b257154380608c11d8c7 Created:20
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: I1001 14:24:07.259391   11828 exec.go:149] Status 'exec' network plugin output: {"ip": "192.168.1.3", "kind": "PodNetworkStatus", "apiVersion": "v1beta
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: , <nil>
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: I1001 14:24:07.261488   11828 kubelet.go:2462] pod waiting > 0, pending
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: E1001 14:24:07.261722   11828 pod_workers.go:112] Error syncing pod bd529a96-6882-11e5-9b85-0800273209ca, skipping: exit status 1
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: I1001 14:24:07.262183   11828 server.go:660] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"busybox", UID:"bd529a96-6882-11e5-9b85-08
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: I1001 14:24:07.270336   11828 status_manager.go:199] Status for pod "busybox_default" updated successfully
Oct 01 14:24:07 cd4-node-1 kubelet[11828]: I1001 14:24:07.272000   11828 config.go:257] Setting pods for source api
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.642037   11828 kubelet.go:1924] SyncLoop (periodic sync)
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.642682   11828 kubelet.go:1888] SyncLoop (housekeeping)
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.649633   11828 docker.go:368] Docker Container: /calico-node is not managed by kubelet.
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.649772   11828 kubelet.go:2529] Generating status for "busybox_default"
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.664808   11828 docker.go:368] Docker Container: /calico-node is not managed by kubelet.
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.666875   11828 manager.go:309] Container inspect result: {ID:8ff588a7a7559041d43ed570b800304d29284573b3c8b257154380608c11d8c7 Created:20
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.760358   11828 docker.go:368] Docker Container: /calico-node is not managed by kubelet.
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.851098   11828 exec.go:149] Status 'exec' network plugin output: {"ip": "192.168.1.3", "kind": "PodNetworkStatus", "apiVersion": "v1beta
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: , <nil>
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.852981   11828 kubelet.go:2462] pod waiting > 0, pending
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.853209   11828 manager.go:1558] Syncing Pod &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:busybox GenerateName: Namespace:default Self
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.853849   11828 manager.go:1569] Found pod infra container for "busybox_default"
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.855037   11828 manager.go:1582] Pod infra container looks good, keep it "busybox_default"
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.855247   11828 manager.go:1597] Container {Name:busybox Image:busybox Command:[sleep 3600] Args:[] WorkingDir: Ports:[] Env:[] Resources
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.855454   11828 manager.go:1688] Got container changes for pod "busybox_default": {StartInfraContainer:false InfraContainerId:8ff588a7a75
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.855648   11828 manager.go:1760] Creating container &{Name:busybox Image:busybox Command:[sleep 3600] Args:[] WorkingDir: Ports:[] Env:[]
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.856956   11828 manager.go:681] Container default/busybox/busybox: setting entrypoint "[sleep 3600]" and command "[]"
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.858634   11828 server.go:660] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"busybox", UID:"bd529a96-6882-11e5-9b85-08
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.878622   11828 docker.go:368] Docker Container: /calico-node is not managed by kubelet.
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.882184   11828 server.go:660] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"busybox", UID:"bd529a96-6882-11e5-9b85-08
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.910220   11828 factory.go:78] Using factory "docker" for container "/system.slice/docker-430aeb280aa8beeeefaf8304b6ee89fcdaae05235aec53d
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.910465   11828 server.go:660] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"busybox", UID:"bd529a96-6882-11e5-9b85-08
Oct 01 14:24:13 cd4-node-1 kubelet[11828]: I1001 14:24:13.915664   11828 kubelet.go:2529] Generating status for "busybox_default"
```

### PR DESCRIPTION
**Pull request by feiskyer**
Delete infra container if network plugin fails.

This PR should fix #14938

### CODE REVIEW

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -1813,13 +1813,20 @@ func (dm *DockerManager) SyncPod(pod *api.Pod, runningPod kubecontainer.Pod, pod
...
 
 		// Call the networking plugin
-		if err == nil {
-			err = dm.networkPlugin.SetUpPod(pod.Namespace, pod.Name, podInfraContainerID)
-		}
+		err = dm.networkPlugin.SetUpPod(pod.Namespace, pod.Name, podInfraContainerID)
 		if err != nil {
 			glog.Errorf("Failed to create pod infra container: %v; Skipping pod %q", err, podFullName)
+			// Delete infra container
+			deleteError := dm.KillContainerInPod(types.UID(podInfraContainerID), nil, pod)
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: nit: keep the deleteError in the `if` block if you are not going to use 
`if err := dm.KillContainerInPod(types.UID(podInfraContainerID), nil, pod); err != nil {`

**Comment by yujuhong** [comment]
lgtm with one nit.

File: pkg/kubelet/dockertools/manager.go (hunk 1 above)
Comment (feiskyer) [inline pkg/kubelet/dockertools/manager.go]: @yujuhong  Fixed

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by yujuhong** [comment]
Thanks for the fix!

**Comment by feiskyer** [comment]
@yujuhong my pleasure

**Comment by karlkfi** [comment]
The above failure is a compilation failure.

Not sure if that's just this branch or this branch merged with master. Might want to rebase.

```
[16:40:00][Step 4/4] _output/local/go/src/k8s.io/kubernetes/pkg/kubelet/dockertools/manager.go:1841: cannot use "k8s.io/kubernetes/pkg/types".UID(podInfraContainerID) (type "k8s.io/kubernetes/pkg/types".UID) as type container.ContainerID in argument to dm.KillContainerInPod
```

**Comment by yujuhong** [comment]
@karlkfi, thanks for the notice! @feiskyer, could you rebase your PR to fix the ContainerID type. Thanks.

**Comment by feiskyer** [comment]
@yujuhong Rebased. How to trigger shippable?

**Comment by karlkfi** [comment]
The failure was on the Mesosphere CI, not Shippable. The build queue si backed up atm, but it should run in < 30m.

**Comment by feiskyer** [comment]
@karlkfi Why Mesosphere CI results is not at the same position as trais-ci?

**Comment by karlkfi** [comment]
Not sure what you mean by "position". Google didn't want our CI to block merging until they trust it, so it's not updating the PR status yet, just commenting. It's also commenting on the commit instead of the PR because that's how the TeamCity integration works (not sure I can change that without writing my own bot).

**Comment by yujuhong** [comment]
@feiskyer, lgtm, could you squash the commits? Thanks!

**Comment by feiskyer** [comment]
@yujuhong Done.

**Comment by dchen1107** [comment]
Thanks for cleaning up this.

**Pull request merged**